File: src/web/src/render/ElasticEffect.ts
/**
 * Image-space elastic dent effect.
 *
 * Calling {@link dent} triggers a spring-damper animation that radially
 * displaces pixels around the click point.  The effect owns its own 60 fps
 * interval timer and stops automatically once the spring has settled.
 *
 * Call {@link applyToPixels} each frame while {@link isActive} returns true to
 * blend the effect into the rendered geometry layer.
 */
export class ElasticEffect {
  private static readonly STIFFNESS   = 350;
  private static readonly DAMPING     = 22;
  /** Initial inward displacement in pixels. */
  private static readonly INIT_DISP   = 10;
  /** Pixel radius of the influence region. */
  static readonly RADIUS = 60;
  private static readonly STOP_THRESH = 0.3;
  private static readonly TICK_MS     = 16;

  private cx = 0;
  private cy = 0;
  private displacement = 0;
  private velocity     = 0;
  private timerId: ReturnType<typeof setInterval> | null = null;

  constructor(private readonly repaint: () => void) {}

  /** Triggers a dent centred on {@code (x, y)}.  Restarts if already running. */
  dent(x: number, y: number): void {
    this.cx           = x;
    this.cy           = y;
    this.displacement = ElasticEffect.INIT_DISP;
    this.velocity     = 0;
    this.clearTimer();
    this.timerId = setInterval(() => this.tick(), ElasticEffect.TICK_MS);
  }

  isActive(): boolean {
    return this.timerId !== null;
  }

  /**
   * Applies radial displacement from {@code src} into {@code dst} (RGBA Uint8ClampedArray).
   * Pixels outside {@link RADIUS} are copied unchanged.  For interior pixels the backward-
   * mapping reads from {@code src} at a position displaced outward by
   * `displacement × 4t(1−t)²` pixels (cubic profile, t = d/RADIUS).
   */
  applyToPixels(
    src: Uint8ClampedArray,
    dst: Uint8ClampedArray,
    w: number,
    h: number,
  ): void {
    const { cx, cy, displacement } = this;
    const R = ElasticEffect.RADIUS;
    const x0 = Math.max(0,     cx - R);
    const x1 = Math.min(w - 1, cx + R);
    const y0 = Math.max(0,     cy - R);
    const y1 = Math.min(h - 1, cy + R);

    for (let oy = y0; oy <= y1; oy++) {
      for (let ox = x0; ox <= x1; ox++) {
        const ddx = ox - cx;
        const ddy = oy - cy;
        const d   = Math.sqrt(ddx * ddx + ddy * ddy);
        if (d === 0 || d >= R) continue;

        const t          = d / R;
        const mag        = displacement * 4 * t * (1 - t) * (1 - t);
        const sourceDist = d + mag;
        const dstBase    = (oy * w + ox) * 4;

        if (sourceDist <= 0) {
          // Clamped to original pixel.
          const srcBase = dstBase;
          dst[dstBase]     = src[srcBase];
          dst[dstBase + 1] = src[srcBase + 1];
          dst[dstBase + 2] = src[srcBase + 2];
          dst[dstBase + 3] = src[srcBase + 3];
          continue;
        }

        const scale = sourceDist / d;
        const ix    = Math.max(0, Math.min(w - 1, Math.round(cx + ddx * scale)));
        const iy    = Math.max(0, Math.min(h - 1, Math.round(cy + ddy * scale)));
        const srcBase = (iy * w + ix) * 4;

        dst[dstBase]     = src[srcBase];
        dst[dstBase + 1] = src[srcBase + 1];
        dst[dstBase + 2] = src[srcBase + 2];
        dst[dstBase + 3] = src[srcBase + 3];
      }
    }
  }

  private tick(): void {
    const dt  = ElasticEffect.TICK_MS / 1000;
    const acc = -ElasticEffect.STIFFNESS * this.displacement
              - ElasticEffect.DAMPING    * this.velocity;
    this.velocity     += acc * dt;
    this.displacement += this.velocity * dt;

    if (
      Math.abs(this.displacement) < ElasticEffect.STOP_THRESH &&
      Math.abs(this.velocity)     < ElasticEffect.STOP_THRESH
    ) {
      this.displacement = 0;
      this.velocity     = 0;
      this.clearTimer();
    }
    this.repaint();
  }

  private clearTimer(): void {
    if (this.timerId !== null) {
      clearInterval(this.timerId);
      this.timerId = null;
    }
  }
}

File: src/main/java/com/ullgren/modern/simple3d/render/ElasticEffect.java
package com.ullgren.modern.simple3d.render;

import javax.swing.Timer;

/**
 * Image-space elastic dent effect.
 * <p>
 * When {@link #trigger(int, int)} is called the effect applies a radial pixel-displacement to
 * the already-rendered frame via {@link #applyToPixels}. A spring-damper drives the
 * displacement scalar from an initial inward compression (positive) through a small outward
 * overshoot (negative) back to zero, creating the feel of touching a soft, elastic object.
 * <p>
 * The effect owns its own 60 fps Swing {@link Timer} and stops automatically once the spring
 * has settled. It is entirely in screen-space, so it works regardless of mesh vertex density.
 */
public class ElasticEffect implements Effect {

  private static final float STIFFNESS   = 350f;
  private static final float DAMPING     = 22f;
  /** Initial inward displacement in pixels. */
  private static final float INIT_DISP   = 10f;
  /** Pixel radius of the influence region. */
  public static final int RADIUS = 60;
  private static final float STOP_THRESH = 0.3f;
  private static final int   TICK_MS     = 16;

  private int   cx, cy;
  private float displacement;
  private float velocity;

  private Timer timer;

  /**
   * @param repaint callback invoked each tick to trigger a canvas repaint
   */
  public ElasticEffect(Runnable repaint) {
    timer = new Timer(TICK_MS, e -> {
      float dt  = TICK_MS / 1000f;
      float acc = -STIFFNESS * displacement - DAMPING * velocity;
      velocity    += acc * dt;
      displacement += velocity * dt;
      if (Math.abs(displacement) < STOP_THRESH && Math.abs(velocity) < STOP_THRESH) {
        displacement = 0;
        velocity     = 0;
        timer.stop();
      }
      repaint.run();
    });
  }

  /**
   * Triggers a dent centred on the given screen coordinates.
   * Restarts the animation if already in progress.
   */
  @Override
  public void trigger(int x, int y) {
    cx           = x;
    cy           = y;
    displacement = INIT_DISP;
    velocity     = 0;
    timer.stop();
    timer.start();
  }

  /** Returns {@code true} while the spring is still in motion. */
  @Override
  public boolean isActive() {
    return timer.isRunning();
  }

  /**
   * Applies the radial displacement from {@code src} into {@code dst}.
   * Pixels outside {@link #RADIUS} are copied unchanged. For pixels inside the radius the
   * backward-mapping samples from {@code src} at a position displaced outward by
   * {@code displacement × sin(π × d / RADIUS)} pixels, compressing the image toward the
   * centre (positive displacement) or expanding it outward (negative overshoot).
   *
   * @param src source pixel array (original frame, must not be {@code dst})
   * @param dst destination pixel array (written into; same dimensions as {@code src})
   * @param w   image width in pixels
   * @param h   image height in pixels
   */
  @Override
  public void applyToPixels(int[] src, int[] dst, int w, int h) {
    int x0 = Math.max(0,     cx - RADIUS);
    int x1 = Math.min(w - 1, cx + RADIUS);
    int y0 = Math.max(0,     cy - RADIUS);
    int y1 = Math.min(h - 1, cy + RADIUS);

    for (int oy = y0; oy <= y1; oy++) {
      for (int ox = x0; ox <= x1; ox++) {
        float ddx = ox - cx;
        float ddy = oy - cy;
        float d   = (float) Math.sqrt(ddx * ddx + ddy * ddy);
        if (d == 0 || d >= RADIUS) continue;

        float t           = d / RADIUS;              // 0..1
        // Cubic profile: peaks at t=1/3, zero at t=0 and t=1 — tighter round shape.
        float mag         = displacement * 4f * t * (1f - t) * (1f - t);
        float sourceDist  = d + mag;
        if (sourceDist <= 0) {
          // Overshoot puts the sample at or past centre — clamp to original.
          dst[oy * w + ox] = src[oy * w + ox];
          continue;
        }
        float scale = sourceDist / d;
        int   ix    = Math.max(0, Math.min(w - 1, Math.round(cx + ddx * scale)));
        int   iy    = Math.max(0, Math.min(h - 1, Math.round(cy + ddy * scale)));
        dst[oy * w + ox] = src[iy * w + ix];
      }
    }
  }
}
